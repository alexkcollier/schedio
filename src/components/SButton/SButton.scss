.button {
  @include button-base;

  &--size-small {
    @include button-size-small;
  }

  &--size-regular {
    @include button-size-regular;
  }

  &--size-large {
    @include button-size-large;
  }

  &--loading {
    color: transparent !important;
    pointer-events: none;
    position: relative;
    text-shadow: none !important;
    user-select: none;

    .icon {
      filter: none;
      stroke: transparent !important;
    }

    .button__spinner {
      left: 50%;
      padding-top: spacing('half');
      position: absolute;
      top: 0;
      transform: translateX(-50%);
    }
  }

  &--icon-only {
    padding: spacing('half');

    .icon {
      margin: 0;

      & + .icon {
        margin-left: spacing('half');
      }
    }
  }

  &__icon {
    &--left {
      margin-left: -1 * spacing('quarter');
      margin-right: spacing('tight');
    }

    &--right {
      margin-left: spacing('tight');
      margin-right: -1 * spacing('quarter');
    }

    .button:disabled & {
      filter: none;
      stroke: color('night', 'lightest');
    }
  }

  @each $color-name in $button-colors {
    &--color-#{$color-name} {
      @include button-color($color-name);

      &-outlined {
        @include button-outline($color-name);
      }

      &-text {
        @include button-text($color-name);
      }
    }
  }
}
