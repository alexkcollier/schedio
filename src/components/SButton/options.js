export {
  colors,
  allShadeOptions,
  colorNames,
} from '@/utils/component-options'

export const sizes = [
  'small',
  'regular',
  'large',
]

export const types = [
  'outlined',
  'text',
]
