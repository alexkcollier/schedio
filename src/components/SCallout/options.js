export const calloutTypes = ['', 'success', 'warning', 'error']
