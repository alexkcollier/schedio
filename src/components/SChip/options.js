import { props } from '@/assets/styles/tokens/dist/tokens.raw.json'

export const colors = [
  ...Object.values(props).reduce((colorList, { category, name }) => {
    if (category === 'background-color') colorList.add(name.split('-')[0])

    return colorList
  }, new Set()),
].sort()
