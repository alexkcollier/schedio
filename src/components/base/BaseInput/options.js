export const allowed = ['email', 'text', 'password', 'url', 'search', ''].sort()
