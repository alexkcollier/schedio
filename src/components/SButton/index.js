import SButton from './SButton.vue';

export { SButton };

export default { install: Vue => Vue.component(SButton.name, SButton) };
