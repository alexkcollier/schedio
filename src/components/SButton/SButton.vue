<template>
  <button
    v-bind="$attrs"
    :class="classList"
    class="button"
    v-on="$listeners"
  >
    <SIcon
      v-if="iconLeft"
      :icon="iconLeft"
      class="button__icon button__icon--left"
    />

    <slot v-if="!iconOnly" />

    <SIcon
      v-if="iconRight"
      :icon="iconRight"
      class="button__icon button__icon--right"
    />
  </button>
</template>

<script>
import { colors, sizes } from './options.js'

export default {
  name: 'SButton',

  props: {
    buttonColor: {
      type: String,
      default: '',
      validator: value => {
        if (!value || colors.includes(value)) return true

        return console.error(`\`buttonColor\` ${value} not found. Allowed colors: ${[...colors]}`)
      }
    },

    buttonSize: {
      type: String,
      default: '',
      validator: value => {
        if (!value || sizes.includes(value)) return true

        return console.error(`\`buttonSizes\` ${value} not found. Allowed sizes: ${sizes}`)
      }
    },

    iconOnly: {
      type: Boolean,
      default: false
    },

    iconLeft: {
      type: String,
      default: ''
    },

    iconRight: {
      type: String,
      default: ''
    }
  },

  computed: {
    classList() {
      return [
        // Handles storybook default
        this.buttonColor ? `button--color-${this.buttonColor}` : 'button--color-green',
        this.buttonSize && `button--size-${this.buttonSize}`,
        this.iconOnly && `button--icon-only`
      ]
    }
  }
}
</script>
