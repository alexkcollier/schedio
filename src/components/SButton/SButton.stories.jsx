import { SButton } from '@/components/SButton'
import { SCallout } from '@/components/SCallout'
import { colors, colorNames, sizes, types, allShadeOptions } from '@/components/SButton/options'
import { SContainerColumn } from '@/components/SContainerColumn'
import { SContainerRow } from '@/components/SContainerRow'
import { SHeading } from '@/components/SHeading'
import { generateHeading } from '@/utils/stories/render-functions'
import PropList from '@@/docs/components/PropList'
import StoryLink from '@@/docs/components/StoryLink'
import { boolean, select, text } from '@storybook/addon-knobs'
import icons from 'feather-icons/dist/icons.json'

const iconNames = Object.keys(icons)
const iconProp = side => select(`icon-${side}`, ['', ...iconNames], '', 'Optional Props')

export default {
  title: 'Components|Buttons/SButton',

  parameters: {
    jest: 'SButton',
  },
}

export const button = () => ({
  props: {
    color: {
      type: String,
      default: select('color', ['', ...colorNames], '', 'Optional Props'),
    },
    shade: {
      type: String,
      default: select('shade', ['', ...allShadeOptions], '', 'Optional Props'),
    },
    type: {
      type: String,
      default: select('type', ['', ...types], '', 'Optional Props'),
    },
    size: {
      type: String,
      default: select('size', ['', ...sizes], '', 'Optional Props'),
    },
    buttonText: {
      type: String,
      default: text('Button text', 'Sample text', 'Slots'),
    },
    disabled: {
      type: Boolean,
      default: boolean('disabled', false, '$attrs'),
    },
    isLoading: {
      type: Boolean,
      default: boolean('is-loading', false, 'Optional Props'),
    },
    iconOnly: {
      type: Boolean,
      default: boolean('icon-only', false, 'Optional Props'),
    },
    iconLeft: {
      type: String,
      default: iconProp('left'),
    },
    iconRight: {
      type: String,
      default: iconProp('right'),
    },
  },
  render (h) {
    const { buttonText, disabled, ...props } = this.$props

    return (
      <div>
        <SHeading>Buttons</SHeading>
        <p>Buttons are used to make actions immediately visible and easy to click.</p>
        <p>
          Use <StoryLink to="Components/Buttons.SButtonGroup" story="Button Group" /> to group
          multiple buttons.
        </p>
        <p>
          If you want the button to just contain an icon, set the <code>icon-only</code> prop to{' '}
          <code>true</code>.
        </p>

        <SHeading level="2">Example</SHeading>
        <SButton {...{ props }} disabled={disabled}>
          {buttonText}
        </SButton>

        <SContainerRow align="center">
          <SContainerColumn>
            <PropList component={SButton} />
          </SContainerColumn>

          <SContainerColumn md="5">
            <SCallout type="warning">
              If you need to set the button's type (e.g.: <code>type="submit"</code>), use the
              <code>native-type</code> prop.
            </SCallout>
          </SContainerColumn>
        </SContainerRow>
      </div >
    )
  },
})

button.story = {
  name: 'Button',
}

export const coloredButton = () => ({
  render (h) {
    return (
      <div>
        <SHeading>Colored Buttons</SHeading>
        {generateButtons(h)}
      </div>
    )
  },
})

coloredButton.story = {
  name: 'Colored Button',

  parameters: {
    options: {
      showPanel: false,
    },
  },
}

export const outlinedButtons = () => ({
  render (h) {
    return (
      <div>
        <SHeading>Outlined Buttons</SHeading>
        {generateButtons(h, { type: 'outlined' })}
      </div>
    )
  },
})

outlinedButtons.story = {
  name: 'Outlined Buttons',

  parameters: {
    options: {
      showPanel: false,
    },
  },
}

export const textButtons = () => ({
  render (h) {
    return (
      <div>
        <SHeading>Text Buttons</SHeading>
        {generateButtons(h, { type: 'text' })}
      </div>
    )
  },
})

textButtons.story = {
  name: 'Text Buttons',

  parameters: {
    options: {
      showPanel: false,
    },
  },
}

function generateButtons (h, propsData = {}) {
  const buttons = Object.entries(colors).map(([color, shades]) => {
    const rows = Object.keys(shades).map((shade) => {
      const props = {
        ...propsData,
        shade,
        color,
      }

      return h(
        SContainerRow,
        { props: { align: 'end' } },
        [
          h(SContainerColumn, { props: { md: 12, lg: 1 } }, [
            h(
              SHeading,
              {
                props: { level: 3 },
                style: { padding: '0.5em', textTransform: 'capitalize' },
              },
              shade,
            ),
          ]),
          generateButtonColumn(h, props, 'No icons'),
          generateButtonColumn(h, { ...props, iconLeft: randIcon() }, 'Icon left'),
          generateButtonColumn(h, { ...props, iconRight: randIcon() }, 'Icon right'),
          generateButtonColumn(h, { ...props, iconLeft: randIcon(), iconRight: randIcon() }, 'Both'),
        ],
      )
    })

    return [
      generateHeading(h, { level: 2, content: color }),
      rows,
    ]
  })

  return h('div', buttons)
}

function generateButtonColumn (h, props, heading) {
  return h(SContainerColumn, { props: { md: 3, lg: 2 } }, [
    generateHeading(h, { level: 3, content: heading }),
    h(SButton, { props }, 'Button'),
  ])
}

function randIcon () {
  const randIdx = Math.round(Math.random() * iconNames.length)
  return iconNames[randIdx]
}
