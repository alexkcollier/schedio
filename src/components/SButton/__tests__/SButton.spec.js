import * as options from '@/components/SButton/options'
import SButton from '@/components/SButton/SButton.vue'
import { shallowMount } from '@vue/test-utils'
import icons from 'feather-icons/dist/icons.json'

const iconList = Object.keys(icons)

describe('SButton.vue', () => {
  const click = jest.fn()
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => {})
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(SButton, {
      slots: {
        default: 'Button Text'
      },
      listeners: { click }
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toBe('Button Text')
  })

  it('handles clicks', () => {
    wrapper.trigger('click')
    expect(click).toHaveBeenCalled()
  })

  options.colors.forEach(color => {
    it(`can be ${color}`, () => {
      wrapper.setProps({ color })
      expect(wrapper.classes()).toContain(`button--color-${color}`)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  options.colors.forEach(color => {
    it(`can be outlined with ${color}`, () => {
      wrapper.setProps({ color, isOutlined: true })
      expect(wrapper.classes()).toContain(`button--color-${color}-outlined`)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('validates fill color', () => {
    shallowMount(SButton, { propsData: { color: 'not a color' } })
    expect(errorSpy).toBeCalled()
  })

  options.sizes.forEach(size => {
    it(`can be ${size}`, () => {
      wrapper.setProps({ size })
      expect(wrapper.classes()).toContain(`button--size-${size}`)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('validates color', () => {
    shallowMount(SButton, { propsData: { size: 'not a size' } })
    expect(errorSpy).toBeCalled()
  })

  it('can have an icon on the left', () => {
    wrapper.setProps({ iconLeft: iconList[0] })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can have an icon on the right', () => {
    wrapper.setProps({ iconRight: iconList[0] })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can an icon and no text', () => {
    const iconButton = shallowMount(SButton, {
      slots: {
        default: 'Button Text'
      },
      listeners: { click },
      propsData: {
        iconOnly: true,
        iconLeft: iconList[0]
      }
    })

    expect(iconButton.text()).toBeFalsy()
    expect(iconButton.html()).toMatchSnapshot()
  })

  it('can have a loading spinner', () => {
    wrapper.setProps({ isLoading: true })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
