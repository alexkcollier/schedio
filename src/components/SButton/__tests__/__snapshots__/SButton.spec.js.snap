// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SButton.vue can be gold 1`] = `
<button class="button button--color-gold">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be green 1`] = `
<button class="button button--color-green">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be grey 1`] = `
<button class="button button--color-grey">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be ice 1`] = `
<button class="button button--color-ice">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be large 1`] = `
<button class="button button--size-large">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be magenta 1`] = `
<button class="button button--color-magenta">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be night 1`] = `
<button class="button button--color-night">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be outlined with gold 1`] = `
<button class="button button--color-gold-outlined">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be outlined with green 1`] = `
<button class="button button--color-green-outlined">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be outlined with grey 1`] = `
<button class="button button--color-grey-outlined">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be outlined with ice 1`] = `
<button class="button button--color-ice-outlined">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be outlined with magenta 1`] = `
<button class="button button--color-magenta-outlined">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be outlined with night 1`] = `
<button class="button button--color-night-outlined">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be outlined with red 1`] = `
<button class="button button--color-red-outlined">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be outlined with spartan_blue 1`] = `
<button class="button button--color-spartan_blue-outlined">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be outlined with turquoise 1`] = `
<button class="button button--color-turquoise-outlined">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be outlined with white 1`] = `
<button class="button button--color-white-outlined">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be red 1`] = `
<button class="button button--color-red">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be regular 1`] = `
<button class="button button--size-regular">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be small 1`] = `
<button class="button button--size-small">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be spartan_blue 1`] = `
<button class="button button--color-spartan_blue">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be turquoise 1`] = `
<button class="button button--color-turquoise">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can be white 1`] = `
<button class="button button--color-white">
  <!---->
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can have a loading spinner 1`] = `
<button class="button button--loading"><span class="button__spinner"><sspinner></sspinner></span>
  <!----> Button Text
  <!----></button>
`;

exports[`SButton.vue can have an icon and no text 1`] = `
<button aria-label="Button" class="button button--icon-only">
  <!---->
  <sicon icon="activity" class="button__icon button__icon--left"></sicon>
  <!---->
  <!---->
</button>
`;

exports[`SButton.vue can have an icon on the left 1`] = `
<button class="button">
  <!---->
  <sicon icon="activity" class="button__icon button__icon--left"></sicon> Button Text
  <!---->
</button>
`;

exports[`SButton.vue can have an icon on the right 1`] = `
<button class="button">
  <!---->
  <!----> Button Text <sicon icon="activity" class="button__icon button__icon--right"></sicon></button>
`;

exports[`SButton.vue renders correctly 1`] = `
<button class="button">
  <!---->
  <!----> Button Text
  <!----></button>
`;
